scheduler/src/cook/mesos/task.clj

```diff
@@ -0,0 +1,167 @@
+  specific ports to use for every task."
+  [available-resources ports-needed]
+  (mapv (fn [port]
+          {:name "ports"
+           :type :value-ranges
+           :role (role-containing-port available-resources port)
+           :ranges [{:begin port :end port}]})
+        ports-needed))
+
+(defn add-ports-to-task-info
+  "Given a set of tasks and offers that were matched together by Fenzo,
+   assigns the specific ports requested by role to each task.
+   Returns the input tasks, decorated with :ports-resource-messages"
+  [available-resources tasks]
+  (let [available-ports (available-resources "ports")]
+    (map (fn [task]
+           (let [ports (:ports-assigned task)
+                 port-env-vars (into {} (map-indexed (fn [i p] [(str "PORT" i) (str p)])
+                                                     ports))]
+             (-> task
+                 (assoc :ports-resource-messages (take-ports available-ports ports))
+                 (update-in [:command :environment] merge port-env-vars))))
+         tasks)))
+
+(defn add-scalar-resources-to-task-info
+  "Given a set of tasks and offers that were matched together by Fenzo,
+   assigns the specific scalar resource requirements by role to each task.
+   Returns the input tasks, decorated with :scalar-resource-messages"
+  [available-resources tasks]
+  (let [results (reduce (fn [{:keys [remaining-resources handled-tasks]} task]
+                          (let [mem-taken (take-resources remaining-resources "mem" (-> task :resources :mem))
+                                cpus-taken (take-resources remaining-resources "cpus" (-> task :resources :cpus))
+                                all-messages (concat (:mesos-messages mem-taken)
+                                                     (:mesos-messages cpus-taken))
+                                new-task (assoc task :scalar-resource-messages all-messages)]
+
+                            {:remaining-resources {"mem" (:remaining-resources mem-taken)
+                                                   "cpus" (:remaining-resources cpus-taken)}
+                             :handled-tasks (conj handled-tasks new-task)}))
+                        {:remaining-resources available-resources
+                         :handled-tasks []}
+                        tasks)]
+    (:handled-tasks results)))
+
+(defn map->mesos-kv
+  "Converts a normal clojure map to a format sometimes employed in mesos messages.
+  e.g. {:foo :bar}  ->  [{:key :foo :value :bar}]"
+  [m key-name]
+  (mapv (fn [kv] {key-name (key kv) :value (val kv)})
+        (vec m)))
+
+(defn task-info->mesos-message
+  "Given a clojure data structure (based on Cook's internal data format for jobs),
+  which has already been decorated with everything we need to know about
+  a task, return a Mesos message that will actually launch that task"
+  [t]
+  {:name (:name t)
+   :slave-id (:slave-id t)
+   :task-id [{:value (:task-id t)}]
```

Comment:
Hmm, it was sent as a list in a mesomatic examples repo I had looked it... But you're right, the actual protobuf only supports a single value here.  Whether it's a single value or a list with 1 element, Mesomatic produces the exact same results, so I'll switch to single value.